### PR TITLE
units: introduce systemd-hibernate-clear.service that clears stale HibernateLocation EFI variable

### PR DESCRIPTION
**Pull request by YHNdnzj**
Fixes https://github.com/systemd/systemd/issues/32021

### CODE REVIEW

File: src/basic/efivars.c [hunk 1]
```diff
@@ -241,7 +242,7 @@ int efi_set_variable(const char *variable, const void *value, size_t size) {
         if (futimens(fd, (struct timespec[2]) {
                                 { .tv_nsec = UTIME_NOW },
                                 { .tv_nsec = UTIME_NOW }
```
Comment (yuwata) [inline src/basic/efivars.c]: Not a matter, but maybe it is better to define this as `TIMESPEC_NOW` with `static const`.

File: src/hibernate-resume/hibernate-resume.c [hunk 2]
```diff
@@ -109,31 +118,44 @@ static int run(int argc, char *argv[]) {
...
-                if (arg_info.efi)
-                        (void) clear_efi_hibernate_location_and_warn();
         }
 
         if (stat(arg_info.device, &st) < 0)
                 return log_error_errno(errno, "Failed to stat resume device '%s': %m", arg_info.device);
 
         if (!S_ISBLK(st.st_mode))
-                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTBLK),
```
Comment (yuwata) [inline src/hibernate-resume/hibernate-resume.c:170]: Maybe, better to move the change into another commit?

File: src/hibernate-resume/hibernate-resume.c [hunk 3]
```diff
@@ -109,31 +118,44 @@ static int run(int argc, char *argv[]) {
...
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                       "Not running in initrd, refusing to initiate resume from hibernation.");
+        if (in_initrd() && arg_clear_efi) {
+                log_info("Running in initrd, --clear-efi has no effect, exiting.");
                 return 0;
+        }
 
-        if (argc > optind) {
-                arg_info.device = argv[optind];
+        if (argc <= optind || arg_clear_efi) {
```
Comment (yuwata) [inline src/hibernate-resume/hibernate-resume.c]: If `--clear-efi` is specified, extra arguments will be ignored. Please refuse when an extra argument is specified, or at least warn about that.

File: units/systemd-hibernate-resume-clear-efi.service.in [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Description=Clear Stale Hibernate EFI Variable
+Documentation=man:systemd-hibernate-resume-clear-efi.service(8)
+
+ConditionPathExists=/sys/firmware/efi/efivars/HibernateLocation-8cf2644b-4b0b-428f-9387-6d876050dc67
+
+DefaultDependencies=no
+Before=sysinit.target shutdown.target
+Conflicts=shutdown.target
+
+ConditionPathExists=!/etc/initrd-release
```
Comment (yuwata) [inline units/systemd-hibernate-resume-clear-efi.service.in]: Please move this line near the other `ConditionPathExists=`.

File: src/basic/efivars.c (hunk 1 above)
Comment (YHNdnzj) [reply src/basic/efivars.c]: Hmm, this only occurs once in the whole codebase, so I'm not sure?

But I'll make it `const`.

File: src/hibernate-resume/hibernate-resume.c [hunk 5]
```diff
@@ -109,16 +122,29 @@ static int run(int argc, char *argv[]) {
...
+        }
 
         if (argc <= optind) {
                 r = setup_hibernate_info_and_warn();
                 if (r <= 0)
                         return r;
 
                 if (arg_info.efi)
-                        (void) clear_efi_hibernate_location_and_warn();
+                        r = clear_efi_hibernate_location_and_warn();
```
Comment (yuwata) [inline src/hibernate-resume/hibernate-resume.c]: Hm, on `--clear-efi` mode, it is not necessary to parse EFI variable and kernel command line, that is `setup_hibernate_info_and_warn()` above is not necessary. Moreover, I think we should clear the EFI var even if we fail to parse them.

I'd like to call `clear_efi_...()` and return earlier:
```c
static int action_clear_efi(void) {
        if (in_initrd()) {
                log_info(...);
                return 0;
        }

        r = clear_efi_...();
        if (r > 0)
                log_notice(...);
        return r;
}

static int run(...) {
        ...
        umask(0022);

        if (arg_clear_efi)
                return action_clear_efi();

        ...
}
```

File: units/systemd-hibernate-resume-clear-efi.service.in [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ConditionPathExists=/sys/firmware/efi/efivars/HibernateLocation-8cf2644b-4b0b-428f-9387-6d876050dc67
+ConditionPathExists=!/etc/initrd-release
+
+DefaultDependencies=no
+Before=sysinit.target shutdown.target
+Conflicts=shutdown.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
```
Comment (yuwata) [inline units/systemd-hibernate-resume-clear-efi.service.in]: Why `RemainAfterExit=yes`??

File: src/hibernate-resume/hibernate-resume.c (hunk 5 above)
Comment (YHNdnzj) [reply src/hibernate-resume/hibernate-resume.c]: This is intentional. If people boot from USB drives/portable systems/... we really shouldn't erase this? Plus, if the system identifier mismatches the resume won't be attempted in the first place?

File: units/systemd-hibernate-resume-clear-efi.service.in (hunk 6 above)
Comment (YHNdnzj) [reply units/systemd-hibernate-resume-clear-efi.service.in]: Hmm, we have this on things like `systemd-boot-update.service` too. Given that this should really only run once during the whole boot, I think it's kinda reasonable?

File: src/hibernate-resume/hibernate-resume.c (hunk 5 above)
Comment (YHNdnzj) [reply src/hibernate-resume/hibernate-resume.c]: I mean, originally I didn't want to add this bit of automation either (you can tell from my response to https://github.com/systemd/systemd/issues/30395 I guess). I imagined very few people running into this, yet that turned out not to be the case.

IOW, one can always manually delete the var in efivarfs if that's strictly necessary. Every failed resume attempt means the loss of some data, so I don't think we should relax this even further...

File: units/systemd-hibernate-resume-clear-efi.service.in (hunk 6 above)
Comment (yuwata) [reply units/systemd-hibernate-resume-clear-efi.service.in]: I see. Yeah, you are right.

File: src/hibernate-resume/hibernate-resume.c (hunk 5 above)
Comment (yuwata) [reply src/hibernate-resume/hibernate-resume.c]: Hm, yeah, may be. Let's wait for another reviewer.

Comment (YHNdnzj) [reply src/hibernate-resume/hibernate-resume.c]: Hmm, but the parse of kernel cmdline should really go. Will update.

File: src/hibernate-resume/hibernate-resume.c [hunk 7]
```diff
@@ -94,6 +106,30 @@ static int setup_hibernate_info_and_warn(void) {
...
+        assert(arg_clear_efi);
+
+        if (in_initrd()) {
+                log_info("Running in initrd, --clear-efi has no effect, exiting.");
+                return 0;
+        }
+
+        /* Let's insist that the system identifier is verified still. After all if things don't match,
+         * the resume wouldn't get triggered in the first place. We should not erase the var if booted
+         * from LiveCD/portable systems/... */
```
Comment (yuwata) [inline src/hibernate-resume/hibernate-resume.c:116]: After re-reading your reply in the above and this comment, yeah, that sounds reasonable.

**Review by yuwata — APPROVED** [review]

File: src/hibernate-resume/hibernate-resume.c [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #include "terminal-util.h"
 
 static HibernateInfo arg_info = {};
+static bool arg_clear_efi = false;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: hmm, i'd drop the EFI name from this. I'd prefer if this would carry a generic name.

Who knows, maybe there's some other scheme how this is info gets persisted one day (thinking about ARM Macs here, which do not have EFI vars). I think the operation should be named generically, and just clear whatever is set and is appropriate for the platform.

hence, maybe just call this --clear?

File: src/hibernate-resume/hibernate-resume.c [hunk 9]
```diff
@@ -32,6 +33,7 @@ static int help(void) {
                "\n%sInitiate resume from hibernation.%s\n\n"
                "  -h --help            Show this help\n"
                "     --version         Show package version\n"
+               "     --clear-efi       Clear stale HibernateLocation EFI variable and exit\n"
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: Well, the "and exit" is redundant. Of course we exit after executing the command.

And I'd drop the reference to the name of the EFI variable. That smells a bit too much of an implementation detail. i.e. sounds great for the man page, but I would not stuff it into the --help text.

And this doesn't just clear stale info, it clears any info afaics, stale or not. Hence, don#t say "stale". I mean, it's hopefully typically used to drop stale info, but that's not really strictly the only case.

Hence maybe "Clear hibernation stroage information from EFI variables" or so?

File: src/hibernate-resume/hibernate-resume.c [hunk 10]
```diff
@@ -94,6 +106,29 @@ static int setup_hibernate_info_and_warn(void) {
...
 
+static int action_clear_efi(void) {
+        int r;
+
+        assert(arg_clear_efi);
+
+        if (in_initrd()) {
+                log_info("Running in initrd, --clear-efi has no effect, exiting.");
+                return 0;
+        }
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: hmm, if people call this from the cmdline i think we should allow this to pass.

Comment (poettering) [reply src/hibernate-resume/hibernate-resume.c]: i mean, key really is that we do not call it from the initrd *automatically*, but calling it explicitly should be fine.

hence this condition i think belongs in the unit file, not in the C code.

(we had similar cases in the past, where we ended up losening the C conditioning in favour of unit file based conditioning)

File: man/systemd-hibernate-resume.service.xml [hunk 11]
```diff
@@ -17,12 +17,14 @@
...
   <refnamediv>
     <refname>systemd-hibernate-resume.service</refname>
+    <refname>systemd-hibernate-resume-clear-efi.service</refname>
     <refname>systemd-hibernate-resume</refname>
     <refpurpose>Resume from hibernation</refpurpose>
   </refnamediv>
 
   <refsynopsisdiv>
     <para><filename>systemd-hibernate-resume.service</filename></para>
+    <para><filename>systemd-hibernate-resume-clear-efi.service</filename></para>
```
Comment (poettering) [inline man/systemd-hibernate-resume.service.xml]: here too, i'd drop the "-efi" suffix, for same reson as before.

Also, there's no resuming here, hence maybe call this just `systemd-hibernate-clear.service`?

**Comment by poettering** [comment]
looks good, just some comments on naming things pretty much

File: src/hibernate-resume/hibernate-resume.c (hunk 9 above)
Comment (YHNdnzj) [reply src/hibernate-resume/hibernate-resume.c]: > Well, the "and exit" is redundant. Of course we exit after executing the command.

Hmm, I see an ambiguity here: people might assume that if and only if this option is specified, the efivar gets cleared before resuming. But it actually just clears it and quits.

**Comment by YHNdnzj** [comment]
Addressed all comments, except for the "and exit" part.
